fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/EvaluateConstantCoalesceRule.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * EvaluateConstantCoalesceRule.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2025 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades.values.simplification;
+
+import com.apple.foundationdb.record.query.plan.cascades.matching.structure.BindingMatcher;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Type;
+import com.apple.foundationdb.record.query.plan.cascades.values.LiteralValue;
+import com.apple.foundationdb.record.query.plan.cascades.values.NullValue;
+import com.apple.foundationdb.record.query.plan.cascades.values.Value;
+import com.apple.foundationdb.record.query.plan.cascades.values.VariadicFunctionValue;
+
+import javax.annotation.Nonnull;
+
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.ValueMatchers.coalesceFunction;
+
+/**
+ * A rule that evaluates {@code Coalesce} with literal {@code 'True}, {@code 'False}, or {@code 'Null} to their
+ * effective coalesced value literal.
+ *
+ * <ul>
+ *     <li>{@code Coalesce(X1, X2, ... Xn, Y ∈ {'True, 'False}, Z1, Z2, .... Zm | Xi ∈ {'Null}, Y ∈ {'True, 'False} Zj ∈ {<ANY VALUE>}) -> Y}</li>
+ * </ul>
+ */
+public class EvaluateConstantCoalesceRule extends ValueSimplificationRule<VariadicFunctionValue> {
+
+    @Nonnull
+    private static final BindingMatcher<VariadicFunctionValue> rootMatcher = coalesceFunction();
+
+    public EvaluateConstantCoalesceRule() {
+        super(rootMatcher);
+    }
+
+    @Override
+    @SuppressWarnings("PMD.AvoidBranchingStatementAsLastInLoop")
+    public void onMatch(@Nonnull final ValueSimplificationRuleCall call) {
+        final var variadicFunctionValue = call.getBindings().get(rootMatcher);
+
+        for (final var child : variadicFunctionValue.getChildren()) {
+            if (canNotFold(child)) {
+                return;
+            }
+            if (child instanceof NullValue) {
+                continue;
+            }
+            call.yieldResult(child);
+            return;
+        }
+    }
+
+    private static boolean canNotFold(@Nonnull final Value value) {
```

Comment:
That should be `cannotFold`. "can not" does not really exist as such.

Reply:
done.